lib/trollop.rb

```diff
@@ -635,170 +575,322 @@ def cloaker(&b)
-  TYPES = FLAG_TYPES + SINGLE_ARG_TYPES + MULTI_ARG_TYPES
-
-  attr_accessor :name, :opts
-
-  def initialize(name, desc="", opts={}, &b)
-    ## fill in :type
-    opts[:type] = # normalize
-      case opts[:type]
-      when :boolean, :bool then :flag
-      when :integer        then :int
-      when :integers       then :ints
-      when :double         then :float
-      when :doubles        then :floats
-      when Class
-        case opts[:type].name
-        when 'TrueClass',
-             'FalseClass'  then :flag
-        when 'String'      then :string
-        when 'Integer'     then :int
-        when 'Float'       then :float
-        when 'IO'          then :io
-        when 'Date'        then :date
-        else
-          raise ArgumentError, "unsupported argument type '#{opts[:type].class.name}'"
-        end
-      when nil             then nil
-      else
-        raise ArgumentError, "unsupported argument type '#{opts[:type]}'" unless TYPES.include?(opts[:type])
-        opts[:type]
-      end
+  attr_accessor :name, :short, :long, :default
+  attr_writer :multi_given
+  
+  def initialize
+    @long = nil
+    @short = nil
+    @name = nil
+    @multi_given = false
+    @hidden = false
+    @default = nil
+    @optshash = Hash.new()
+  end
+  
+  def opts (key)
+    @optshash[key]
+  end
+
+  def opts= (o)
+    @optshash = o
+  end
+  
+  ## Indicates a flag option, which is an option without an argument
+  def flag? ; false ; end
+  def single_arg?
+    !self.multi_arg? and !self.flag?
+  end
 
-    ## for options with :multi => true, an array default doesn't imply
-    ## a multi-valued argument. for that you have to specify a :type
-    ## as well. (this is how we disambiguate an ambiguous situation;
-    ## see the docs for Parser#opt for details.)
-    disambiguated_default = if opts[:multi] && opts[:default].kind_of?(Array) && !opts[:type]
-      opts[:default].first
-    else
-      opts[:default]
+  def multi ; @multi_given ; end
+  alias multi? multi
+
+  ## Indicates that this is a multivalued (Array type) argument
+  def multi_arg? ; false ; end
+  ## note: Option-Types with both multi_arg? and flag? false are single-parameter (normal) options.
+
+  def array_default? ; self.default.kind_of?(Array) ; end
+  
+  def short? ; short && short != :none ; end
+
+  def callback ; opts(:callback) ; end
+  def desc ; opts(:desc) ; end
+
+  def required? ; opts(:required) ; end
+
+  def parse (_paramlist, _neg_given)
+    raise NotImplementedError, "parse must be overridden for newly registered type"
+  end
+
+  # provide type-format string.  default to empty, but user should probably override it
+  def type_format ; "" ; end
+  
+  def educate
+    (short? ? "-#{short}, " : "") + "--#{long}" + type_format + (flag? && default ? ", --no-#{long}" : "")
+  end
+
+  ## Format the educate-line description including the default-value(s)
+  def description_with_default
+    return desc unless default
+    defstring = begin
```

Comment:
this got a little too complex...

```ruby
default_s = case default
            when $stdout   then "<stdout>"
            when $stdin    then "<stdin>"
            when $stderr   then "<stderr>"
            when Array
              default.join(", ")
            else
              default.to_s
            end

if desc =~ /\.$/
  "#{desc} (Default: #{default_s})"
else
  "#{desc} (default: #{default_s})"
end
```

Reply:
This was basically unchanged from before, I just cleaned it up a little more.
It would be nice if we could get rid of the capitalization difference that keys off a period at the end of the description.  I can't think of a compelling reason why that feature would've been added or why it should be kept.